### PR TITLE
Refactor all opcodes to use the constraint builder + other utils

### PR DESCRIPTION
**Pull request by Brechtpd**
Updated all opcodes implementations to use all the utilities we currently have available that did not already do so. Unfortunately this makes for a large diff but better earlier than later I think. The comparator gadget was already updated in #120 so did not update it again here.

All implementations should be identical to the previous version. If you find a change it's very likely I made a mistake in the conversion. I did do a small change in `push.rs` lines 75..91 because I think there was an issue there (did not check the MSB selector) but maybe I misunderstood something.

### CODE REVIEW

**Review by miha-stopar — APPROVED** [review]
LGTM!

**Comment by miha-stopar** [comment]
> LGTM!

Just saw there is a conflict with math_gadgets.rs, can you resolve it @Brechtpd? Will merge right after that.

**Comment by Brechtpd** [comment]
> > LGTM!
> 
> Just saw there is a conflict with math_gadgets.rs, can you resolve it @Brechtpd? Will merge right after that.

Done! @miha-stopar

**Pull request merged**
